persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/DelegatingAuthorizer.java

```diff
@@ -0,0 +1,222 @@
+
+import com.datastax.bdp.cassandra.auth.AuthRequestExecutionException;
+import io.stargate.auth.AuthorizationOutcome;
+import io.stargate.auth.AuthorizationProcessor;
+import io.stargate.auth.PermissionKind;
+import io.stargate.auth.entity.AccessPermission;
+import io.stargate.auth.entity.AuthorizedResource;
+import io.stargate.auth.entity.EntitySelector;
+import io.stargate.auth.entity.ImmutableAccessPermission;
+import io.stargate.auth.entity.ImmutableActor;
+import io.stargate.auth.entity.ImmutableAuthorizedResource;
+import io.stargate.auth.entity.ResourceKind;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import org.apache.cassandra.auth.AuthenticatedUser;
+import org.apache.cassandra.auth.CassandraAuthorizer;
+import org.apache.cassandra.auth.DataResource;
+import org.apache.cassandra.auth.FunctionResource;
+import org.apache.cassandra.auth.GrantMode;
+import org.apache.cassandra.auth.IResource;
+import org.apache.cassandra.auth.Permission;
+import org.apache.cassandra.auth.RoleResource;
+import org.apache.cassandra.auth.permission.CorePermission;
+import org.apache.cassandra.stargate.exceptions.RequestValidationException;
+import org.javatuples.Pair;
+
+public class DelegatingAuthorizer extends CassandraAuthorizer {
+
+  private AuthorizationProcessor authProcessor;
+
+  public void setProcessor(AuthorizationProcessor processor) {
+    authProcessor = processor;
+  }
+
+  @Override
+  public Set<Permission> grant(
+      AuthenticatedUser performer,
+      Set<Permission> permissions,
+      IResource resource,
+      RoleResource grantee,
+      GrantMode... grantModes) {
+    AuthorizationProcessor processor = authProcessor;
+    if (processor == null) {
+      return super.grant(performer, permissions, resource, grantee, grantModes);
+    }
+
+    Collection<CompletableFuture<Void>> stages = new ArrayList<>(grantModes.length);
+
+    for (GrantMode grantMode : grantModes) {
+      Pair<PermissionKind, AuthorizationOutcome> mode = grantMode(grantMode);
+      CompletionStage<Void> stage =
+          processor.addPermissions(
+              ImmutableActor.of(performer.getName()),
+              mode.getValue1(),
+              mode.getValue0(),
+              permissions(permissions),
+              resource(resource),
+              role(grantee));
+
+      stages.add(stage.toCompletableFuture());
+    }
+
+    get(CompletableFuture.allOf(stages.toArray(new CompletableFuture[0])));
+
+    // assume all permissions were granted if we did not an exception
+    return permissions;
+  }
+
+  @Override
+  public Set<Permission> revoke(
+      AuthenticatedUser performer,
+      Set<Permission> permissions,
+      IResource resource,
+      RoleResource revokee,
+      GrantMode... grantModes) {
+    AuthorizationProcessor processor = authProcessor;
```

Comment:
why do you need this reassignment, instead of using `authProcessor` directly?

Reply:
it's a remnant of previous experiment when the field was an `AtomicReference`... I'll clean up.